src/main/java/org/ngafid/uploads/process/format/ParquetFileProcessor.java

```diff
@@ -0,0 +1,249 @@
+import org.ngafid.uploads.process.FatalFlightFileException;
+import org.ngafid.uploads.process.FlightMeta;
+
+import java.io.IOException;
+import java.sql.Connection;
+import java.util.*;
+import java.util.logging.Logger;
+import java.util.stream.Stream;
+
+/**
+ * Processes a Parquet file and extracts flights.
+ */
+public class ParquetFileProcessor {
+    private static final Logger LOG = Logger.getLogger(ParquetFileProcessor.class.getName());
+    private final Connection connection;
+    private final InputFile inputFile;
+    private final String filename;
+
+    public ParquetFileProcessor(Connection connection, InputFile inputFile, String filename) {
+        this.connection = connection;
+        this.inputFile = inputFile;
+        this.filename = filename;
+    }
+
+    public Stream<FlightBuilder> parse() {
+        LOG.info("Parsing Parquet file: " + filename);
+        List<FlightBuilder> flightBuilders = new ArrayList<>();
+
+        try (ParquetReader<GenericRecord> reader = AvroParquetReader.<GenericRecord>builder(inputFile).build()) {
+            GenericRecord record;
+
+           int counter = 0;
+            while ((record = reader.read()) != null) {
+                FlightMeta flightMeta = new FlightMeta();
+
+                Log.info("Processing metadata for flight " + counter);
+                processMetaData(record, flightMeta);
+
+                Map<String, DoubleTimeSeries> doubleTimeSeries = new HashMap<>();
+                Map<String, StringTimeSeries> stringTimeSeries = new HashMap<>();
+
+                populateTimeSeries(record, doubleTimeSeries, stringTimeSeries);
+
+                int numberRows = doubleTimeSeries.get(Parameters.UNIX_TIME_SECONDS).size();
+                LOG.info("Number of rows for flight " + counter + ": " + numberRows);
+
+
+                flightBuilders.add(new FlightBuilder(flightMeta, doubleTimeSeries, stringTimeSeries));
+                counter++;
+
+                // How many flights we are processing (testing)
+                  if (counter > 10) break;
+
+            }
+        } catch (IOException e) {
+            LOG.severe("Error reading Parquet file: " + e.getMessage());
+            throw new RuntimeException("Failed to parse Parquet file", e);
+        } catch (FatalFlightFileException e) {
+            throw new RuntimeException(e);
+        }
+        return flightBuilders.stream();
+    }
+
+    private void populateTimeSeries(GenericRecord record, Map<String, DoubleTimeSeries> doubleTimeSeries, Map<String, StringTimeSeries> stringTimeSeries) {
+
+        // Collect data for each time series
+        List<Double> timeValues = new ArrayList<>();
+        List<Double> latitudeValues = new ArrayList<>();
+        List<Double> longitudeValues = new ArrayList<>();
+        List<Double> speedValues = new ArrayList<>();
+        List<Double> accelerationValues = new ArrayList<>();
+        List<Double> altitudeValues = new ArrayList<>();
```

Comment:
Avoid using `List<Double>` or any generic primitive  -- it can be slow and consumes more memory. Generics using primitives in java unfortunately are Boxed, meaning they are references to objects that contain values rather than values.

Reply:
`DoubleTimeSeries` was actually written by me to avoid this exact problem. It can be used like a list (it has add and will resize itself when necessary).